sway-core/src/ir_generation/const_eval.rs

```diff
@@ -650,29 +720,84 @@ fn const_eval_typed_expr(
                 _ => {
                     return Err(ConstEvalError::CannotBeEvaluatedToConst {
                         span: expr.span.clone(),
-                    })
+                    });
                 }
             }
         }
         ty::TyExpressionVariant::Ref(_) | ty::TyExpressionVariant::Deref(_) => {
             return Err(ConstEvalError::CompileError);
         }
-        ty::TyExpressionVariant::Reassignment(_)
-        | ty::TyExpressionVariant::FunctionParameter
+        ty::TyExpressionVariant::EnumTag { exp } => {
+            let value = const_eval_typed_expr(lookup, known_consts, exp, allow_configurables)?
+                .map(|x| x.value);
+            if let Some(ConstantValue::Struct(fields)) = value {
+                Some(fields[0].clone())
+            } else {
+                return Err(ConstEvalError::CompileError);
+            }
+        }
+        ty::TyExpressionVariant::UnsafeDowncast { exp, .. } => {
+            let value = const_eval_typed_expr(lookup, known_consts, exp, allow_configurables)?
+                .map(|x| x.value);
+            if let Some(ConstantValue::Struct(fields)) = value {
+                Some(fields[1].clone())
+            } else {
+                return Err(ConstEvalError::CompileError);
+            }
+        }
+        ty::TyExpressionVariant::WhileLoop {
```

Comment:
Fair point.

We do need loops, because we need to be able to encode arrays, for example.
I will put a limit of 1M iterations at the moment.

We may even put a time-based limit/warning later.